File: firmware/src/sign_matcher.c
// firmware/src/sign_matcher.c
//
// C port of core/sign_matcher.py — the core sign matching algorithm.
// Linear error-to-PWM mapping with dead-band clamping.

#include "sign_matcher.h"
#include <math.h>
#include <string.h>

// ─── Constants ──────────────────────────────────────────────────────

#define MAX_ERROR_FOR_FULL_PWM_DEG 40.0f
#define DEFAULT_TOLERANCE_DEG       8.0f
#define DEFAULT_IMU_TOLERANCE_DEG  15.0f

// ─── Sign Definition ────────────────────────────────────────────────

typedef struct {
    int8_t  sign_id;
    float   target_angles[5];    // degrees
    float   roll_deg;
    float   pitch_deg;
    float   yaw_deg;
    float   tolerance_band;
    float   imu_tolerance;
    bool    is_loaded;
} SignDef_t;

static SignDef_t sign_library[MAX_SIGNS];

// ─── ISL Fingerspelling Angles (hardcoded for now) ──────────────────
// In production, load from LittleFS JSON files.
// This matches the Python ISL_FINGERSPELLING_ANGLES dictionary.

static const float ISL_ANGLES[26][5] = {
    {60, 90, 90, 90, 90},   // A
    {90,  0,  0,  0,  0},   // B
    {30, 30, 30, 30, 30},   // C
    {90,  0, 90, 90, 90},   // D
    {60, 60, 60, 60, 60},   // E
    { 0, 90,  0,  0,  0},   // F
    { 0,  0, 90, 90, 90},   // G
    {90,  0,  0, 90, 90},   // H
    {90, 90, 90, 90,  0},   // I
    {90, 90, 90, 90,  0},   // J
    { 0,  0,  0, 90, 90},   // K
    { 0,  0, 90, 90, 90},   // L
    {90, 90, 90, 90, 90},   // M
    {90, 90, 90, 90, 90},   // N
    {20, 20, 20, 20, 20},   // O
    { 0,  0, 90, 90, 90},   // P
    { 0,  0, 90, 90, 90},   // Q
    {90,  0,  0, 90, 90},   // R
    {90, 90, 90, 90, 90},   // S
    {30, 90, 90, 90, 90},   // T
    {90,  0,  0, 90, 90},   // U
    {90,  0,  0, 90, 90},   // V
    {90,  0,  0,  0, 90},   // W
    {90, 45, 90, 90, 90},   // X
    { 0, 90, 90, 90,  0},   // Y
    {90,  0, 90, 90, 90},   // Z
};

static const float ISL_IMU[26][3] = {
    // {roll, pitch, yaw}
    {0,   0,  90},  // A
    {0,   0,  90},  // B
    {0,  10,  90},  // C
    {0,   0,  90},  // D
    {0,   0,  90},  // E
    {0,   0,  90},  // F
    {0,  90,   0},  // G
    {0,  90,   0},  // H
    {0,   0,  90},  // I
    {0,   0,  90},  // J
    {0,   0,  90},  // K
    {0,   0,  90},  // L
    {0, -20,  90},  // M
    {0,  20,  90},  // N
    {0,   0,  90},  // O
    {0, -90,   0},  // P
    {0,  90,   0},  // Q
    {0,   0,  90},  // R
    {0,   0,  90},  // S
    {0,   0,  90},  // T
    {0,   0,  90},  // U
    {0,   0,  90},  // V
    {0,   0,  90},  // W
    {0,   0,  90},  // X
    {0,   0,  90},  // Y
    {0,   0,  90},  // Z
};

// ─── Initialisation ─────────────────────────────────────────────────

void sign_matcher_init(void) {
    memset(sign_library, 0, sizeof(sign_library));

    for (int i = 0; i < MAX_SIGNS; i++) {
        sign_library[i].sign_id = (int8_t)i;
        memcpy(sign_library[i].target_angles, ISL_ANGLES[i], sizeof(float) * 5);
        sign_library[i].roll_deg  = ISL_IMU[i][0];
        sign_library[i].pitch_deg = ISL_IMU[i][1];
        sign_library[i].yaw_deg   = ISL_IMU[i][2];
        sign_library[i].tolerance_band = DEFAULT_TOLERANCE_DEG;
        sign_library[i].imu_tolerance  = DEFAULT_IMU_TOLERANCE_DEG;
        sign_library[i].is_loaded = true;
    }
}

// ─── Match Computation ──────────────────────────────────────────────

void sign_matcher_compute(
    int8_t sign_id,
    const float flex_angles[5],
    float imu_roll,
    float imu_pitch,
    float imu_yaw,
    uint8_t motor_duty[5],
    bool *all_correct,
    uint8_t *wrist_pattern
) {
    if (sign_id < 0 || sign_id >= MAX_SIGNS || !sign_library[sign_id].is_loaded) {
        memset(motor_duty, 0, 5);
        *all_correct = false;
        *wrist_pattern = 0;
        return;
    }

    SignDef_t *sign = &sign_library[sign_id];
    float tolerance = sign->tolerance_band;
    bool fingers_correct = true;

    // Per-finger error → PWM duty (linear with dead-band)
    for (int i = 0; i < 5; i++) {
        float err = fabsf(flex_angles[i] - sign->target_angles[i]);

        if (err <= tolerance) {
            motor_duty[i] = 0;  // dead band
        } else if (err >= MAX_ERROR_FOR_FULL_PWM_DEG) {
            motor_duty[i] = 255;
            fingers_correct = false;
        } else {
            motor_duty[i] = (uint8_t)(
                (err - tolerance) / (MAX_ERROR_FOR_FULL_PWM_DEG - tolerance) * 255.0f
            );
            fingers_correct = false;
        }
    }

    // IMU orientation error
    float imu_err_roll  = fabsf(imu_roll  - sign->roll_deg);
    float imu_err_pitch = fabsf(imu_pitch - sign->pitch_deg);
    float imu_err_yaw   = fabsf(imu_yaw   - sign->yaw_deg);
    bool wrist_ok = (imu_err_roll  <= sign->imu_tolerance) &&
                    (imu_err_pitch <= sign->imu_tolerance) &&
                    (imu_err_yaw   <= sign->imu_tolerance);

    *all_correct = fingers_correct && wrist_ok;
    *wrist_pattern = wrist_ok ? 0 : 1;
}

File: core/sign_matcher.py
# core/sign_matcher.py
#
# Core sign matching engine — the heart of the glove's feedback logic.
# Fully unit-testable without hardware.

import os
import glob
import json
import numpy as np
from dataclasses import dataclass
from typing import Optional


@dataclass
class MatchResult:
    """Result of comparing current hand pose against a target ISL sign."""

    is_match: bool                  # True if ALL fingers within tolerance
    finger_errors: list             # degrees error per finger [5 floats]
    imu_errors: dict                # {'roll': float, 'pitch': float, 'yaw': float}
    motor_duties: list              # PWM 0–255 per finger [5 ints]
    wrist_needs_correction: bool    # True if IMU error > imu_tolerance
    all_correct: bool               # True if match + IMU correct
    error_magnitude: float          # RMS of finger errors (useful for logging)


class SignMatcher:
    """
    Core sign matching algorithm.

    Error-to-PWM mapping: LINEAR, not logarithmic.

    Why linear? The publication's haptic effectiveness claim requires the
    vibration intensity to be proportional to the angular error. A logarithmic
    map would compress large errors and the gradient would be weak for users
    making coarse corrections. A linear map gives maximum gradient signal
    throughout the correction range.

    Max error for full PWM: 40° (configurable). At 40° error → duty=255.
    Dead band: 8° → duty=0 (configurable per sign).
    Between 8° and 40°: linear interpolation.
    """

    MAX_ERROR_FOR_FULL_PWM_DEG = 40.0

    def __init__(self, sign_library_path: str = "data/signs/"):
        self.signs: dict[int, dict] = {}
        self._load_library(sign_library_path)
        self.active_sign_id: Optional[int] = None

    def _load_library(self, path: str):
        """Load all ISL sign JSON files from the given directory."""
        for filepath in glob.glob(os.path.join(path, "*.json")):
            with open(filepath) as f:
                sign = json.load(f)
                self.signs[sign["sign_id"]] = sign
        print(f"SignMatcher: loaded {len(self.signs)} signs.")

    def set_target_sign(self, sign_id: int):
        """Set the active target sign for matching."""
        if sign_id not in self.signs:
            raise ValueError(f"Sign ID {sign_id} not in library (available: {sorted(self.signs.keys())})")
        self.active_sign_id = sign_id

    def get_sign_label(self, sign_id: int) -> str:
        """Get the human-readable label for a sign ID."""
        if sign_id in self.signs:
            return self.signs[sign_id].get("label", f"SIGN_{sign_id}")
        return f"UNKNOWN_{sign_id}"

    def compute_match(self, current_angles: list, current_imu: dict) -> Optional[MatchResult]:
        """
        Primary matching computation. Call this every 10 ms (signMatchTask period).

        Parameters
        ----------
        current_angles : list of float
            [thumb, index, middle, ring, pinky] in degrees.
        current_imu : dict
            {'roll': float, 'pitch': float, 'yaw': float} in degrees.

        Returns
        -------
        MatchResult or None
            None if no target sign is active.
        """
        if self.active_sign_id is None:
            return None

        sign = self.signs[self.active_sign_id]
        target = np.array(sign["target_angles"])
        current = np.array(current_angles)
        tolerance = sign["tolerance_band"]

        # Per-finger absolute error
        finger_errors = np.abs(current - target).tolist()

        # IMU orientation error
        wrist = sign["wrist_orientation"]
        imu_errors = {
            "roll":  abs(current_imu["roll"]  - wrist["roll_deg"]),
            "pitch": abs(current_imu["pitch"] - wrist["pitch_deg"]),
            "yaw":   abs(current_imu["yaw"]   - wrist["yaw_deg"]),
        }
        imu_tolerance = sign.get("imu_tolerance", 15.0)
        wrist_needs_correction = any(e > imu_tolerance for e in imu_errors.values())

        # Error-to-PWM mapping: linear with dead-band clamping
        motor_duties = []
        for err in finger_errors:
            if err <= tolerance:
                duty = 0  # dead band → silence
            elif err >= self.MAX_ERROR_FOR_FULL_PWM_DEG:
                duty = 255
            else:
                # Linear interpolation: tolerance → 0, MAX_ERROR → 255
                duty = int(
                    (err - tolerance)
                    / (self.MAX_ERROR_FOR_FULL_PWM_DEG - tolerance)
                    * 255
                )
            motor_duties.append(duty)

        is_match = all(e <= tolerance for e in finger_errors)
        all_correct = is_match and not wrist_needs_correction
        error_magnitude = float(np.sqrt(np.mean(np.array(finger_errors) ** 2)))

        return MatchResult(
            is_match=is_match,
            finger_errors=finger_errors,
            imu_errors=imu_errors,
            motor_duties=motor_duties,
            wrist_needs_correction=wrist_needs_correction,
            all_correct=all_correct,
            error_magnitude=error_magnitude,
        )
